node/src/components/linear_chain/state.rs

```diff
@@ -367,3 +368,141 @@ impl LinearChain {
         }
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use crate::{crypto::generate_ed25519_keypair, logging, testing::TestRng};
+    use casper_types::EraId;
+
+    use super::*;
+
+    #[test]
+    fn new_block_no_sigs() {
+        let mut rng = TestRng::new();
+        let protocol_version = ProtocolVersion::V1_0_0;
+        let mut lc = LinearChain::new(protocol_version, 1u64, 1u64);
+        let block = Block::random(&mut rng);
+        let execution_results = HashMap::new();
+        let new_block_outcomes =
+            lc.handle_new_block(Box::new(block.clone()), execution_results.clone());
+        let block_hash = *block.hash();
+        match &*new_block_outcomes {
+            [Outcome::StoreBlock(outcome_block), Outcome::StoreExecutionResults(outcome_block_hash, outcome_execution_results)] =>
+            {
+                assert_eq!(&**outcome_block, &block);
+                assert_eq!(outcome_block_hash, &block_hash);
+                assert_eq!(outcome_execution_results, &execution_results);
+            }
+            others => panic!("unexpected outcome: {:?}", others),
+        }
+        let block_stored_outcomes = lc.handle_put_block(Box::new(block.clone()));
+        match &*block_stored_outcomes {
+            [Outcome::AnnounceBlock(announced_block)] => {
+                assert_eq!(&**announced_block, &block);
+            }
+            others => panic!("unexpected outcome: {:?}", others),
+        }
+        assert_eq!(
+            lc.latest_block(),
+            &Some(block),
+            "should update the latest block"
+        );
+    }
+
+    #[test]
+    fn pending_sig_rejected() {
+        let mut rng = TestRng::new();
+        let protocol_version = ProtocolVersion::V1_0_0;
+        let mut lc = LinearChain::new(protocol_version, 1u64, 1u64);
+        let block_hash = BlockHash::random(&mut rng);
+        let valid_sig = FinalitySignature::random_for_block(block_hash, 0);
+        let handle_sig_outcomes = lc.handle_finality_signature(Box::new(valid_sig.clone()), false);
+        assert_eq!(handle_sig_outcomes.len(), 1);
```

Comment:
Why not check that it's the expected outcome? (`LoadSignatures`?)

Reply:
I think I couldn't get the compiler to work as I wanted but then learned how to do it later. I will change that.

Reply:
[`6f5f46c` (#1284)](https://github.com/CasperLabs/casper-node/pull/1284/commits/6f5f46c35c509dfb3f246460187696ebb353bc9b)